pkg/backup/backupschedule/backup_schedule_manager.go

```diff
@@ -288,26 +356,260 @@ func (bm *backupScheduleManager) backupGCByMaxReservedTime(bs *v1alpha1.BackupSc
+	expiredBackups, err = calExpiredSnapshotBackupsWithLogBackup(backupsList, expiredTSO)
+	if err != nil {
+		return nil, 0, perrors.Annotate(err, "calculate expired backups which should be deleted")
+	}
+
+	// no expired backup should be deleted, we can skip and wait next time.
+	// because log backup checkpoint ts is always changing, but we should't frequently delete or truncate.
+	// we can delete or truncate log backup when there has backup should be deleted, and this is not frequent and will not trigger repeatedly.
+	if len(expiredBackups) == 0 {
+		return nil, 0, nil
+	}
+
+	// calculate truncate tso
+	truncateTSO, err = calLogBackupExpiredTSO(backupsList, logBackup, expiredTSO)
+	if err != nil {
+		return nil, 0, perrors.Annotate(err, "calculate expired log backup tso which should be truncated")
+	}
+
+	return expiredBackups, truncateTSO, nil
+}
+
+func caculateExpiredBackups(backupsList []*v1alpha1.Backup, reservedTime time.Duration) ([]*v1alpha1.Backup, error) {
+	expiredTS := config.TSToTSO(time.Now().Add(-1 * reservedTime).Unix())
+	i := 0
+	for ; i < len(backupsList); i++ {
+		startTS, err := config.ParseTSString(backupsList[i].Status.CommitTs)
+		if err != nil {
+			return nil, perrors.Annotatef(err, "parse start tso: %s", backupsList[i].Status.CommitTs)
+		}
+		if startTS >= expiredTS {
+			break
+		}
+	}
+	return backupsList[:i], nil
+}
+
+// calculateLatestTSO calculate the latest tso in auto backups and log backups
+func calculateLatestTSO(backupsList []*v1alpha1.Backup, logBackup *v1alpha1.Backup) (uint64, error) {
+	var (
+		currentBackupTS    uint64
+		latestBackupTS     uint64
+		latestCheckPointTS uint64
+		latestTS           uint64
+		err                error
+	)
+
+	// get latestCheckPointTs
+	latestCheckPointTS, err = config.ParseTSString(logBackup.Status.LogCheckpointTs)
+	if err != nil {
+		return 0, perrors.Annotatef(err, "parse checkpoint ts of log backup %s/%s", logBackup.Namespace, logBackup.Name)
+	}
+
+	for _, backup := range backupsList {
+		currentBackupTS, err = config.ParseTSString(backup.Status.CommitTs)
+		if err != nil {
+			return 0, perrors.Annotatef(err, "parse backup ts of backup %s/%s", backup.Namespace, backup.Name)
+		}
+		if currentBackupTS > latestBackupTS {
+			latestBackupTS = currentBackupTS
+		}
+	}
+
+	// use the larger of backup and logBackup ts as the latestTS
+	latestTS = latestBackupTS
+	if latestCheckPointTS > latestTS {
+		latestTS = latestCheckPointTS
+	}
+
+	return latestTS, nil
+}
+
+// calculateExpiredTSO calculate latestTSO - reservedTime
+func calculateExpiredTSO(latestTSO uint64, reservedTime time.Duration) uint64 {
+	latestTS := config.TSOToTS(latestTSO)
+	expiredTS := time.Unix(latestTS, 0).Add(-1 * reservedTime).Unix()
+	return config.TSToTSO(expiredTS)
+}
+
+// calExpiredSnapshotBackupsWithLogBackup according to expired tso calculate expired backups
+func calExpiredSnapshotBackupsWithLogBackup(backupsList []*v1alpha1.Backup, expiredTSO uint64) ([]*v1alpha1.Backup, error) {
```

Comment:
wrong naming?

Reply:
ok, change it to be `calExpiredBackupsWithLogBackupOn` and add some note.